DSEGOG-119 DSEGOG-120 Data export

kaperoo opened the pull request:
## Description
This PR build on #302 and adds the export dialogue and the actual export functionality. 
It also (temporarily) fixes DSEGOG-141 by sorting by timestamp in ascending order if no other sort exists.

New _Export_ button next to session save buttons:
![image](https://github.com/ral-facilities/operationsgateway/assets/72079517/325d9631-76bf-4118-bbf3-f8192834e420)


The export dialogue with all the export options:
![image](https://github.com/ral-facilities/operationsgateway/assets/72079517/8733f369-cf44-46ad-9e56-3e4dd45cd902)
(Currently exporting _Selected Rows_ with no rows selected just throws an error in the console).

Note that the record selection is saved in the session.

## Testing instructions
Run the updated unit and e2e tests.
Also, added new testcases: `export.test.tsx` and `export.cy.ts`.

- [ ] Review code
- [ ] Check Actions build
- [ ] Review changes to test coverage

## Agile board tracking
connect to DSEGOG-119

kaperoo committed:
export dialog and slice

kaperoo committed:
updated snapshots

kaperoo committed:
fix unit tests

kaperoo committed:
move export button to a separate component

kaperoo committed:
fix e2e test

kaperoo committed:
initial tests for export button and dialogue

kaperoo committed:
name changes + initial e2e tests

kaperoo committed:
better coverage for export button

kaperoo committed:
use the /export endpoint

kaperoo committed:
update export dialogue unit tests

kaperoo committed:
initial useExportData unit tests

kaperoo committed:
updated unit tests and correct download name

kaperoo committed:
updated tests + selection bug fix

kaperoo committed:
fix failing unit tests

kaperoo committed:
selection tests in recordTable

kaperoo committed:
tidy-up the code

louise-davies commented on cypress/e2e/export.cy.ts:
```diff
@@ -0,0 +1,88 @@
+    cy.findByRole('dialog').should('exist');
+    cy.findByRole('button', { name: 'Cancel' }).click();
+    cy.findByRole('dialog').should('not.exist');
+  });
```
This can be one "can open and close the dialog" test since the first test case is covered in the latter anyway

louise-davies commented on src/api/export.tsx:
```diff
@@ -0,0 +1,183 @@
+      };
+      const startIndex =
+        exportType === 'Visible Rows' ? page * resultsPerPage : 0;
+      const stopIndex =
+        exportType === 'Visible Rows'
+          ? startIndex + resultsPerPage
+          : exportType === 'All Rows'
+            ? maxShots
+            : 0;
```
This logic doesn't work when maxShots is Unlimited i.e. the value of `maxShots` is `Infinity` - the backend doesn't like `Infinity` as a value. We should probably not set `offsetParams` at all in that case

louise-davies replied:
We should also probably have a test case to cover Unlimited shots since handling with `Infinity` is weirder than the difference between 50 and 1000

louise-davies commented on src/export/exportButton.component.test.tsx:
```diff
@@ -0,0 +1,25 @@
+    fireEvent.click(exportButton);
+    const exportDialogue = screen.getByRole('dialog');
+    expect(exportDialogue).toBeInTheDocument();
+  });
```
This test seems to be covered by the previous test case

louise-davies commented on src/export/exportButton.component.test.tsx:
```diff
@@ -0,0 +1,25 @@
+  it('opens and closes the dialog on button clicks', async () => {
+    renderComponentWithProviders(<ExportButton />);
+
+    fireEvent.click(screen.getByText('Export'));
```
Is there a reason `fireEvent` is needed here over `userEvent`?

louise-davies commented on src/export/exportDialogue.component.tsx:
```diff
@@ -0,0 +1,110 @@
+const ExportDialogue = (props: ExportDialogueProps) => {
+  const { open, onClose } = props;
+
+  const mutate = useExportData().mutate;
```
Style-wise, I think `const { mutate } = useExportData();` looks better, plus is more extensible if we ever need other variables from `useExportData`

louise-davies commented on src/export/exportDialogue.component.tsx:
```diff
@@ -0,0 +1,110 @@
+
+  const handleClose = React.useCallback(() => {
+    onClose();
+  }, [onClose]);
```
This wrapper function isn't needed as `handleClose` as the exact same call signature as `onClose` - just pass `onClose` directly where you pass `handleClose`

louise-davies commented on src/views/recordTable.component.test.tsx:
```diff
@@ -99,6 +99,61 @@ describe('Record Table', () => {
+        rest.get('/records', (req, res, ctx) => {
+          return res(ctx.status(200), ctx.json(recordsJson.slice(0, 3)));
+        })
+      );
```
Why is this necessary? I get in the previous test we're reducing the number of results to make asserting all rows selected easier, but since in this test we're just indexing individual rows I would guess the test works fine without this

louise-davies reviewed (changes requested):
Overall looking good - just found 1 bug to do with Unlimited shots and various code style comments.

Another comment is that if the export request takes a long time - there's no visual to the user that the request is processing. Perhaps we can use a [LoadingButton](https://mui.com/material-ui/react-button/#loading-button)? Would also need to disable clicking away from the export dialogue when fetching as well... Alternatively, eCat has a "generating data" overlay which we could also do with a [Backdrop](https://mui.com/material-ui/react-backdrop/)

![image](https://github.com/ral-facilities/operationsgateway/assets/22525471/fe670b83-e3e3-44e7-af75-bb5b4121c635)

kaperoo committed:
address PR comments + export backdrop

kaperoo committed:
update/fix tests

kaperoo committed:
temp fix for default sorting bug

kaperoo committed:
updated records tests

kaperoo commented:
Addressed comments, added a _"generating export data"_ backdrop and fixed the default timestamp sorting issue.

|Backdrop in light mode:|Backdrop in dark mode:|
|:-------------------------:|:-------------------------:|
|<img width="250" alt="light" src="https://github.com/ral-facilities/operationsgateway/assets/72079517/f4025b7b-83e1-4177-8248-310cb1332cf5">|<img width="250" alt="dark" src="https://github.com/ral-facilities/operationsgateway/assets/72079517/5b2f7986-a846-4761-b11d-03730a0be6d5">|

louise-davies reviewed (approved):
Seems to be working well - all rows export is working with Unlimited shots mode and the loading dialogue looks good :)
